clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -205,6 +233,24 @@ public void send(String method, Object... args) throws Exception {
         transport.send(message);
     }
 
+    /**
+     * Invokes a hub method on the server using the specified method name.
+     * Does not wait for a response from the receiver.
+     * @param method The name of the server method to invoke.
+     * @param args The arguments to be passed to the method.
+     * @throws Exception If there was an error while sending.
+     */
+    public CompletableFuture sendAsync(String method, Object... args) throws Exception {
```

Comment:
We shouldn't actually return `CompletableFuture`, we should return [`CompletionStage`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html).

`CompletionStage` is what consumers need to chain other code to run when the Future completes. `CompletableFuture` gives the caller the ability to "complete" the task, which isn't their job :).

From what I can tell:

* `CompletableFuture` (Java) <==> `TaskCompletionSource` (.NET)
* `CompletionStage` (Java) <==> `Task` (.NET)

`CompletableFuture` just implements `CompletionStage` so we can just return the future instance, I just think we should change the type of the return value.

The one thing that gets slightly harder is synchronously blocking since `CompletionStage` doesn't have the `.get` method. However, it does have `.toCompletableFuture()` so if you want to be synchronous you can call `.sendAsync().toCompletableFuture().get()`

Reply:
That also strongly discourages people from using `.get()` which is the equivalent of `.Wait()` or `.Result` on `Task`, both of which should be avoided in .NET.

Reply:
Looking at some Azure SDKs, it looks like just returning `CompletableFuture` is the common pattern. For example, the [Azure Service Bus SDK for Java](https://github.com/Azure/azure-service-bus-java/blob/dev/azure-servicebus/src/main/java/com/microsoft/azure/servicebus/QueueClient.java#L176).

So I guess just go with what they do.